// mjml/lib/components/Section.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _class, _class2, _temp;

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _MJMLElement = require('./decorators/MJMLElement');

var _MJMLElement2 = _interopRequireDefault(_MJMLElement);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Sections are intended to be used as rows within your email. They will be used to structure the layout.
 */
var Section = (_dec = (0, _MJMLElement2.default)({
  tagName: 'mj-section',
  attributes: {
    'background-repeat': 'repeat',
    'padding': '20px 0',
    'background-size': 'auto'
  }
}), _dec(_class = (_temp = _class2 = function (_Component) {
  _inherits(Section, _Component);

  function Section() {
    _classCallCheck(this, Section);

    return _possibleConstructorReturn(this, Object.getPrototypeOf(Section).apply(this, arguments));
  }

  _createClass(Section, [{
    key: 'isFullWidth',
    value: function isFullWidth() {
      var mjAttribute = this.props.mjAttribute;

      return mjAttribute('full-width') == 'full-width';
    }
  }, {
    key: 'getStyles',
    value: function getStyles() {
      var mjAttribute = this.props.mjAttribute;

      var background = mjAttribute('background-url') ? {
        background: 'url(' + mjAttribute('background-url') + ') top center / ' + (mjAttribute('background-size') || '') + ' ' + (mjAttribute('background-repeat') || '')
      } : {
        background: mjAttribute('background-color')
      };

      return _lodash2.default.merge({}, this.constructor.baseStyles, {
        td: {
          fontSize: 0,
          textAlign: mjAttribute('text-align'),
          verticalAlign: mjAttribute('vertical-align'),
          padding: mjAttribute('padding'),
          paddingTop: mjAttribute('padding-top'),
          paddingBottom: mjAttribute('padding-bottom'),
          paddingRight: mjAttribute('padding-right'),
          paddingLeft: mjAttribute('padding-left')
        },
        div: {
          maxWidth: mjAttribute('parentWidth')
        }
      }, {
        div: this.isFullWidth() ? {} : _lodash2.default.cloneDeep(background),
        table: this.isFullWidth() ? {} : _lodash2.default.cloneDeep(background),
        tableFullwidth: this.isFullWidth() ? _lodash2.default.cloneDeep(background) : {}
      });
    }
  }, {
    key: 'renderFullWidthSection',
    value: function renderFullWidthSection() {
      var mjAttribute = this.props.mjAttribute;

      return _react2.default.createElement(
        'table',
        { 'data-legacy-background': mjAttribute('background-url'),
          border: '0',
          cellPadding: '0',
          cellSpacing: '0',
          'data-width': mjAttribute('parentWidth'),
          style: _lodash2.default.merge({}, this.styles.tableFullwidth, this.styles.table) },
        _react2.default.createElement(
          'tbody',
          null,
          _react2.default.createElement(
            'tr',
            null,
            _react2.default.createElement(
              'td',
              null,
              this.renderSection()
            )
          )
        )
      );
    }
  }, {
    key: 'renderSection',
    value: function renderSection() {
      var _props = this.props;
      var renderWrappedOutlookChildren = _props.renderWrappedOutlookChildren;
      var mjAttribute = _props.mjAttribute;

      var fullWidth = this.isFullWidth();

      return _react2.default.createElement(
        'div',
        { style: this.styles.div },
        _react2.default.createElement(
          'table',
          { className: 'outlook-background-fix-open',
            'data-url': mjAttribute('background-url') || '',
            'data-legacy-background': fullWidth ? undefined : mjAttribute('background-url'),
            border: '0',
            cellPadding: '0',
            cellSpacing: '0',
            'data-legacy-align': 'center',
            'data-width': mjAttribute('parentWidth'),
            style: this.styles.table },
          _react2.default.createElement(
            'tbody',
            null,
            _react2.default.createElement(
              'tr',
              null,
              _react2.default.createElement(
                'td',
                { style: this.styles.td },
                renderWrappedOutlookChildren()
              )
            )
          )
        )
      );
    }
  }, {
    key: 'render',
    value: function render() {
      this.styles = this.getStyles();

      return this.isFullWidth() ? this.renderFullWidthSection() : this.renderSection();
    }
  }]);

  return Section;
}(_react.Component), _class2.baseStyles = {
  div: {
    margin: "0 auto"
  },
  table: {
    width: "100%",
    fontSize: "0"
  },
  td: {
    textAlign: 'center',
    verticalAlign: 'top'
  }
}, _temp)) || _class);
exports.default = Section;
